File: random-search.py
import csv
import random
import os
import statistics
import time
from datetime import datetime

import numpy as np
import matplotlib.pyplot as plt

# ── Load Data from Files ──────────────────────────────────────────────────────

PROBLEM_SIZE = "10x10"
DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), f"podaci-{PROBLEM_SIZE}")


def load_matrix_csv(filepath):
    matrix = []
    with open(filepath, "r") as file:
        reader = csv.reader(file)
        next(reader)
        for row in reader:
            matrix.append([int(value) for value in row[1:]])
    return matrix


def load_demand_csv(filepath):
    demand = []
    with open(filepath, "r") as file:
        reader = csv.reader(file)
        for row in reader:
            if len(row) >= 2:
                try:
                    demand.append(int(row[1]))
                except ValueError:
                    continue
    return demand


def load_matrix_tsv(filepath):
    matrix = []
    with open(filepath, "r") as file:
        lines = file.readlines()
        for line in lines[1:]:
            parts = line.strip().split("\t")
            if len(parts) > 1:
                row = [int(value) for value in parts[1:]]
                matrix.append(row)
    return matrix


def load_demand_tsv(filepath):
    demand = []
    with open(filepath, "r") as file:
        for line in file.readlines():
            parts = line.strip().split("\t")
            if len(parts) >= 2:
                try:
                    demand.append(int(parts[1]))
                except ValueError:
                    continue
    return demand


def load_data(problem_size, data_directory):
    size_number = problem_size.split("x")[0]

    csv_time = os.path.join(data_directory, f"vremena_izvrsavanja_{size_number}.csv")
    csv_profit = os.path.join(data_directory, f"zarada_po_servisu_{size_number}.csv")
    csv_demand = os.path.join(data_directory, f"zahtevi_za_servisima_{size_number}.csv")

    if os.path.exists(csv_time):
        time_matrix = load_matrix_csv(csv_time)
        profit_matrix = load_matrix_csv(csv_profit)
        demand_vector = load_demand_csv(csv_demand)
    else:
        time_matrix = load_matrix_tsv(os.path.join(data_directory, "vremena_izvrsavanja.txt"))
        profit_matrix = load_matrix_tsv(os.path.join(data_directory, "zarada_po_servisu.txt"))
        demand_vector = load_demand_tsv(os.path.join(data_directory, "zahtevi_za_servisima.txt"))

    return time_matrix, profit_matrix, demand_vector


print(f"Ucitavanje podataka za {PROBLEM_SIZE}...")
_time_list, _profit_list, _demand_list = load_data(PROBLEM_SIZE, DATA_DIR)

TIME = np.array(_time_list, dtype=np.int32)
PROFIT = np.array(_profit_list, dtype=np.int32)
DEMAND = np.array(_demand_list, dtype=np.int32)

NUM_SERVICES, NUM_COMPUTERS = TIME.shape
MAX_TIME = 2880

print(f"Dimenzije problema: {NUM_SERVICES} servisa x {NUM_COMPUTERS} racunara")

# ── Random Search Hyperparameters (mirrors GA budget) ─────────────────────────

POPULATION_SIZE = 100
GENERATIONS = 500

print("\nRandom Search Parametri:")
print(f"  Velicina populacije: {POPULATION_SIZE}")
print(f"  Broj generacija: {GENERATIONS}")
print(f"  Ukupno evaluacija po pokretanju: {POPULATION_SIZE * GENERATIONS}")

# ── Precomputed Helpers ───────────────────────────────────────────────────────

PROFIT_PER_MINUTE = PROFIT.astype(np.float64) / TIME.astype(np.float64)
MAX_UNITS = (MAX_TIME // TIME).astype(np.int32)

SERVICES_BY_RATIO_PER_COMPUTER = [
    np.argsort(PROFIT_PER_MINUTE[:, computer]).tolist()
    for computer in range(NUM_COMPUTERS)
]

SERVICES_BEST_FIRST_PER_COMPUTER = [
    np.argsort(PROFIT_PER_MINUTE[:, computer])[::-1].tolist()
    for computer in range(NUM_COMPUTERS)
]


# ── Repair ────────────────────────────────────────────────────────────────────

def repair(chromosome):
    np.clip(chromosome, 0, None, out=chromosome)

    row_totals = chromosome.sum(axis=1)
    over_demand = row_totals > DEMAND
    if over_demand.any():
        ratios = DEMAND[over_demand].astype(np.float64) / row_totals[over_demand]
        chromosome[over_demand] = (chromosome[over_demand] * ratios[:, np.newaxis]).astype(np.int32)

    for computer in range(NUM_COMPUTERS):
        used = int(np.sum(TIME[:, computer] * chromosome[:, computer]))
        if used > MAX_TIME:
            for service in SERVICES_BY_RATIO_PER_COMPUTER[computer]:
                if used <= MAX_TIME:
                    break
                if chromosome[service, computer] > 0:
                    excess = used - MAX_TIME
                    units_to_remove = min(int(chromosome[service, computer]), (excess + int(TIME[service, computer]) - 1) // int(TIME[service, computer]))
                    chromosome[service, computer] -= units_to_remove
                    used -= units_to_remove * int(TIME[service, computer])

    for computer in range(NUM_COMPUTERS):
        used = int(np.sum(TIME[:, computer] * chromosome[:, computer]))
        remaining = MAX_TIME - used
        for service in SERVICES_BEST_FIRST_PER_COMPUTER[computer]:
            if remaining < TIME[service, computer]:
                continue
            demand_left = int(DEMAND[service]) - int(chromosome[service].sum())
            if demand_left <= 0:
                continue
            add = min(demand_left, remaining // int(TIME[service, computer]))
            if add > 0:
                chromosome[service, computer] += add
                remaining -= add * int(TIME[service, computer])
            if remaining < 5:
                break

    return chromosome


# ── Fitness ───────────────────────────────────────────────────────────────────

def fitness(chromosome):
    return int(np.sum(PROFIT * chromosome))


# ── Initialization ────────────────────────────────────────────────────────────

def init_random():
    upper_bounds = np.minimum(DEMAND[:, np.newaxis], MAX_UNITS)
    chromosome = np.zeros((NUM_SERVICES, NUM_COMPUTERS), dtype=np.int32)
    for service in range(NUM_SERVICES):
        for computer in range(NUM_COMPUTERS):
            chromosome[service, computer] = random.randint(0, int(upper_bounds[service, computer]))
    return repair(chromosome)


def init_population(size):
    return [init_random() for _ in range(size)]


# ── Main Random Search Loop ───────────────────────────────────────────────────

def run_random_search():
    best_ever = None
    best_fitness_ever = 0
    generation_best_history = []

    print("Pokretanje random pretrage...\n")
    for generation in range(GENERATIONS):
        population = init_population(POPULATION_SIZE)
        fitnesses = [fitness(chromosome) for chromosome in population]

        generation_best_index = max(range(len(fitnesses)), key=lambda idx: fitnesses[idx])
        generation_best_fitness = fitnesses[generation_best_index]
        generation_best_history.append(generation_best_fitness)

        if generation_best_fitness > best_fitness_ever:
            best_fitness_ever = generation_best_fitness
            best_ever = population[generation_best_index].copy()

        if generation % 10 == 0:
            print(f"Gen {generation:4d} | Gen: {generation_best_fitness:,} | Best ever: {best_fitness_ever:,}")

    return best_ever, best_fitness_ever, generation_best_history


# ── Output ────────────────────────────────────────────────────────────────────

def print_solution(chromosome, total_profit):
    print("\n" + "=" * 80)
    print(f"  MAKSIMALNA ZARADA: {total_profit:,} dinara")
    print("=" * 80)

    print("\n── Iskoriscenje racunara ────────────────────────────────────────────────")
    time_per_computer = np.sum(TIME * chromosome, axis=0)
    total_used = int(time_per_computer.sum())
    total_capacity = NUM_COMPUTERS * MAX_TIME
    for computer in range(NUM_COMPUTERS):
        used = int(time_per_computer[computer])
        percent = 100 * used / MAX_TIME
        if percent < 99.0:
            print(f"  C{computer+1:3d}: {used:5d}/{MAX_TIME} min  ({percent:5.1f}%)")
    avg_utilization = 100 * total_used / total_capacity
    print(f"\n  Prosecna iskoriscenje: {avg_utilization:.2f}%")

    print("\n── Iskoriscenje traznje servisa ─────────────────────────────────────────")
    allocated_per_service = chromosome.sum(axis=1)
    services_at_max = int(np.sum(allocated_per_service == DEMAND))
    services_unused = int(np.sum(allocated_per_service == 0))
    print(f"  Servisi sa 100% iskoriscenjem: {services_at_max}/{NUM_SERVICES}")
    print(f"  Nekorisceni servisi: {services_unused}/{NUM_SERVICES}")

    print("=" * 80)


# ── Multi-Run ─────────────────────────────────────────────────────────────────

NUM_RUNS = 15

RESULTS_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), f"rezultati-{PROBLEM_SIZE}")
os.makedirs(RESULTS_DIR, exist_ok=True)


def save_results(run_profits, run_times, best_overall_profit):
    median_profit = statistics.median(run_profits)
    timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    filename = f"random-{POPULATION_SIZE}-{GENERATIONS}-{NUM_RUNS}runs-{timestamp}.txt"
    filepath = os.path.join(RESULTS_DIR, filename)

    with open(filepath, "w") as file:
        file.write(f"Random Search Parametri:\n")
        file.write(f"  Velicina populacije: {POPULATION_SIZE}\n")
        file.write(f"  Broj generacija: {GENERATIONS}\n")
        file.write(f"  Ukupno evaluacija po pokretanju: {POPULATION_SIZE * GENERATIONS}\n")
        file.write(f"  Broj pokretanja: {NUM_RUNS}\n")
        file.write(f"\n{'='*80}\n")
        file.write(f"  REZULTATI SVIH POKRETANJA\n")
        file.write(f"{'='*80}\n\n")

        for run_index, (profit, elapsed) in enumerate(zip(run_profits, run_times)):
            file.write(f"  Pokretanje {run_index + 1:3d}: {profit:,} dinara  ({elapsed:.2f}s)\n")

        file.write(f"\n{'='*80}\n")
        file.write(f"  MEDIJANA: {median_profit:,.1f} dinara\n")
        file.write(f"  NAJBOLJI: {best_overall_profit:,} dinara\n")
        file.write(f"  NAJGORI:  {min(run_profits):,} dinara\n")
        total_time = sum(run_times)
        file.write(f"\n  UKUPNO VREME:  {total_time:.2f}s\n")
        file.write(f"  PROSECNO VREME: {statistics.mean(run_times):.2f}s\n")
        file.write(f"{'='*80}\n")

    print(f"\nRezultati sacuvani u: {filepath}")
    return filepath


def save_convergence_plot(all_histories):
    timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    filename = f"random-{POPULATION_SIZE}-{GENERATIONS}-{NUM_RUNS}runs-{timestamp}.png"
    filepath = os.path.join(RESULTS_DIR, filename)

    plt.figure(figsize=(10, 6))
    for run_index, history in enumerate(all_histories):
        running_best = np.maximum.accumulate(history)
        x_values = [(generation + 1) * POPULATION_SIZE for generation in range(len(history))]
        plt.plot(x_values, running_best, linewidth=1, alpha=0.7, label=f"Pokretanje {run_index + 1}")

    plt.xlabel("Broj generacija x velicina populacije (broj evaluacija)")
    plt.ylabel("Vrednost funkcije cilja (zarada)")
    plt.title(f"Konvergencija Random Search — {PROBLEM_SIZE}, pop={POPULATION_SIZE}, gen={GENERATIONS}")
    plt.grid(True, alpha=0.3)
    plt.legend(loc="lower right", fontsize=8)
    plt.tight_layout()
    plt.savefig(filepath, dpi=120)
    plt.close()
    print(f"Grafik sacuvan u: {filepath}")
    return filepath


# ── Entry Point ───────────────────────────────────────────────────────────────

if __name__ == "__main__":
    run_profits = []
    run_times = []
    all_histories = []
    best_overall_chromosome = None
    best_overall_profit = 0

    for run_index in range(NUM_RUNS):
        print(f"\n{'#'*80}")
        print(f"  POKRETANJE {run_index + 1}/{NUM_RUNS}")
        print(f"{'#'*80}")

        random.seed(run_index)
        np.random.seed(run_index)
        start_time = time.time()
        best_chromosome, best_profit, generation_best_history = run_random_search()
        all_histories.append(generation_best_history)
        elapsed_time = time.time() - start_time
        run_profits.append(best_profit)
        run_times.append(elapsed_time)

        if best_profit > best_overall_profit:
            best_overall_profit = best_profit
            best_overall_chromosome = best_chromosome

        print(f"\n  Pokretanje {run_index + 1} zavrseno | Zarada: {best_profit:,} | Vreme: {elapsed_time:.2f}s")

    median_profit = statistics.median(run_profits)

    print(f"\n{'='*80}")
    print(f"  SUMARNI REZULTATI ({NUM_RUNS} pokretanja)")
    print(f"{'='*80}")
    for run_index, (profit, elapsed) in enumerate(zip(run_profits, run_times)):
        print(f"  Pokretanje {run_index + 1:3d}: {profit:,} dinara  ({elapsed:.2f}s)")
    print(f"\n  MEDIJANA: {median_profit:,.1f} dinara")
    print(f"  NAJBOLJI: {best_overall_profit:,} dinara")
    print(f"  NAJGORI:  {min(run_profits):,} dinara")
    print(f"  PROSEK:   {statistics.mean(run_profits):,.1f} dinara")
    if len(run_profits) > 1:
        print(f"  STD DEV:  {statistics.stdev(run_profits):,.1f} dinara")
    total_time = sum(run_times)
    print(f"\n  UKUPNO VREME:  {total_time:.2f}s")
    print(f"  PROSECNO VREME: {statistics.mean(run_times):.2f}s")
    print(f"{'='*80}")

    print_solution(best_overall_chromosome, best_overall_profit)
    save_results(run_profits, run_times, best_overall_profit)
    save_convergence_plot(all_histories)

File: random-search.cpp
// Random Search — C++ port of random-search.py
// Baseline comparison for genetic-algorithm.cpp.
// Build: g++ -O3 -march=native -flto -std=c++17 random-search.cpp -o random-search
// Run:   ./random-search

#include <algorithm>
#include <chrono>
#include <climits>
#include <cmath>
#include <cstdint>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <random>
#include <sstream>
#include <string>
#include <vector>

namespace fs = std::filesystem;

// ── Parameters ───────────────────────────────────────────────────────────────
static const std::string PROBLEM_SIZE = "1000x1000";
static const long long EVALUATIONS = 2000000; // total random solutions per run
static const int SAMPLE_INTERVAL = 20000;     // record best-so-far every N evaluations
static const int NUM_RUNS = 1;
static const int MAX_TIME = 2880;

// ── Globals (problem data) ───────────────────────────────────────────────────
static int NUM_SERVICES = 0;
static int NUM_COMPUTERS = 0;
static std::vector<std::vector<int>> TIME_MAT;
static std::vector<std::vector<int>> PROFIT_MAT;
static std::vector<int> DEMAND;
static std::vector<std::vector<int>> MAX_UNITS;
static std::vector<std::vector<double>> PROFIT_PER_MINUTE;
static std::vector<std::vector<int>> SERVICES_BY_RATIO_PER_COMPUTER;   // worst first
static std::vector<std::vector<int>> SERVICES_BEST_FIRST_PER_COMPUTER; // best first

using Gene = int16_t;
using Chromosome = std::vector<Gene>;

inline Gene &cell(Chromosome &c, int s, int k) { return c[s * NUM_COMPUTERS + k]; }
inline Gene cell(const Chromosome &c, int s, int k) { return c[s * NUM_COMPUTERS + k]; }

// ── RNG ──────────────────────────────────────────────────────────────────────
static std::mt19937 rng(std::random_device{}());
inline int randint(int low, int high) { return std::uniform_int_distribution<int>(low, high)(rng); }

// ── Data loading ─────────────────────────────────────────────────────────────
static std::vector<std::vector<int>> load_matrix(const std::string &path, char delim, bool skip_header)
{
    std::vector<std::vector<int>> matrix;
    std::ifstream file(path);
    if (!file)
    {
        std::cerr << "Cannot open " << path << "\n";
        std::exit(1);
    }
    std::string line;
    if (skip_header)
        std::getline(file, line);
    while (std::getline(file, line))
    {
        std::stringstream ss(line);
        std::string cell_str;
        std::vector<int> row;
        bool first = true;
        while (std::getline(ss, cell_str, delim))
        {
            if (first)
            {
                first = false;
                continue;
            }
            try
            {
                row.push_back(std::stoi(cell_str));
            }
            catch (...)
            {
            }
        }
        if (!row.empty())
            matrix.push_back(row);
    }
    return matrix;
}

static std::vector<int> load_demand(const std::string &path, char delim, bool skip_header)
{
    std::vector<int> demand;
    std::ifstream file(path);
    if (!file)
    {
        std::cerr << "Cannot open " << path << "\n";
        std::exit(1);
    }
    std::string line;
    if (skip_header)
        std::getline(file, line);
    while (std::getline(file, line))
    {
        std::stringstream ss(line);
        std::string label, value_str;
        if (std::getline(ss, label, delim) && std::getline(ss, value_str, delim))
        {
            try
            {
                demand.push_back(std::stoi(value_str));
            }
            catch (...)
            {
            }
        }
    }
    return demand;
}

static void load_data()
{
    fs::path base = fs::path(__FILE__).parent_path();
    fs::path dir = base / ("podaci-" + PROBLEM_SIZE);

    std::string size_number = PROBLEM_SIZE.substr(0, PROBLEM_SIZE.find('x'));
    fs::path csv_time = dir / ("vremena_izvrsavanja_" + size_number + ".csv");
    fs::path csv_profit = dir / ("zarada_po_servisu_" + size_number + ".csv");
    fs::path csv_demand = dir / ("zahtevi_za_servisima_" + size_number + ".csv");

    if (fs::exists(csv_time))
    {
        TIME_MAT = load_matrix(csv_time.string(), ',', true);
        PROFIT_MAT = load_matrix(csv_profit.string(), ',', true);
        DEMAND = load_demand(csv_demand.string(), ',', false);
    }
    else
    {
        TIME_MAT = load_matrix((dir / "vremena_izvrsavanja.txt").string(), '\t', true);
        PROFIT_MAT = load_matrix((dir / "zarada_po_servisu.txt").string(), '\t', true);
        DEMAND = load_demand((dir / "zahtevi_za_servisima.txt").string(), '\t', false);
    }

    NUM_SERVICES = (int)TIME_MAT.size();
    NUM_COMPUTERS = (int)TIME_MAT[0].size();

    MAX_UNITS.assign(NUM_SERVICES, std::vector<int>(NUM_COMPUTERS));
    PROFIT_PER_MINUTE.assign(NUM_SERVICES, std::vector<double>(NUM_COMPUTERS));
    for (int s = 0; s < NUM_SERVICES; ++s)
        for (int k = 0; k < NUM_COMPUTERS; ++k)
        {
            MAX_UNITS[s][k] = MAX_TIME / TIME_MAT[s][k];
            PROFIT_PER_MINUTE[s][k] = (double)PROFIT_MAT[s][k] / (double)TIME_MAT[s][k];
        }

    SERVICES_BY_RATIO_PER_COMPUTER.assign(NUM_COMPUTERS, {});
    SERVICES_BEST_FIRST_PER_COMPUTER.assign(NUM_COMPUTERS, {});
    for (int k = 0; k < NUM_COMPUTERS; ++k)
    {
        std::vector<int> idx(NUM_SERVICES);
        std::iota(idx.begin(), idx.end(), 0);
        std::sort(idx.begin(), idx.end(), [k](int a, int b)
                  { return PROFIT_PER_MINUTE[a][k] < PROFIT_PER_MINUTE[b][k]; });
        SERVICES_BY_RATIO_PER_COMPUTER[k] = idx;
        std::reverse(idx.begin(), idx.end());
        SERVICES_BEST_FIRST_PER_COMPUTER[k] = idx;
    }
}

// ── Repair ───────────────────────────────────────────────────────────────────
static void repair(Chromosome &chromosome)
{
    for (auto &gene : chromosome)
        if (gene < 0)
            gene = 0;

    for (int s = 0; s < NUM_SERVICES; ++s)
    {
        long long total = 0;
        for (int k = 0; k < NUM_COMPUTERS; ++k)
            total += cell(chromosome, s, k);
        if (total > DEMAND[s])
        {
            double ratio = (double)DEMAND[s] / (double)total;
            for (int k = 0; k < NUM_COMPUTERS; ++k)
                cell(chromosome, s, k) = (Gene)(cell(chromosome, s, k) * ratio);
        }
    }

    for (int k = 0; k < NUM_COMPUTERS; ++k)
    {
        long long used = 0;
        for (int s = 0; s < NUM_SERVICES; ++s)
            used += (long long)TIME_MAT[s][k] * cell(chromosome, s, k);
        if (used > MAX_TIME)
        {
            for (int s : SERVICES_BY_RATIO_PER_COMPUTER[k])
            {
                if (used <= MAX_TIME)
                    break;
                if (cell(chromosome, s, k) > 0)
                {
                    long long excess = used - MAX_TIME;
                    int t = TIME_MAT[s][k];
                    int units_to_remove = (int)std::min<long long>(cell(chromosome, s, k), (excess + t - 1) / t);
                    cell(chromosome, s, k) = (Gene)(cell(chromosome, s, k) - units_to_remove);
                    used -= (long long)units_to_remove * t;
                }
            }
        }
    }

    for (int k = 0; k < NUM_COMPUTERS; ++k)
    {
        long long used = 0;
        for (int s = 0; s < NUM_SERVICES; ++s)
            used += (long long)TIME_MAT[s][k] * cell(chromosome, s, k);
        long long remaining = MAX_TIME - used;
        for (int s : SERVICES_BEST_FIRST_PER_COMPUTER[k])
        {
            int t = TIME_MAT[s][k];
            if (remaining < t)
                continue;
            long long row_sum = 0;
            for (int kk = 0; kk < NUM_COMPUTERS; ++kk)
                row_sum += cell(chromosome, s, kk);
            long long demand_left = DEMAND[s] - row_sum;
            if (demand_left <= 0)
                continue;
            int add = (int)std::min<long long>(demand_left, remaining / t);
            if (add > 0)
            {
                cell(chromosome, s, k) = (Gene)(cell(chromosome, s, k) + add);
                remaining -= (long long)add * t;
            }
            if (remaining < 5)
                break;
        }
    }
}

// ── Fitness ──────────────────────────────────────────────────────────────────
static long long fitness(const Chromosome &chromosome)
{
    long long total = 0;
    for (int s = 0; s < NUM_SERVICES; ++s)
        for (int k = 0; k < NUM_COMPUTERS; ++k)
            total += (long long)PROFIT_MAT[s][k] * cell(chromosome, s, k);
    return total;
}

// ── Init ─────────────────────────────────────────────────────────────────────
static Chromosome init_random()
{
    Chromosome chromosome(NUM_SERVICES * NUM_COMPUTERS, 0);
    for (int s = 0; s < NUM_SERVICES; ++s)
    {
        for (int k = 0; k < NUM_COMPUTERS; ++k)
        {
            int upper = std::min(DEMAND[s], MAX_UNITS[s][k]);
            cell(chromosome, s, k) = (Gene)randint(0, upper);
        }
    }
    repair(chromosome);
    return chromosome;
}

// ── Random Search Loop ───────────────────────────────────────────────────────
struct SamplePoint
{
    long long evaluations;
    long long best_fitness; // best-so-far at this evaluation count
};

struct RunResult
{
    Chromosome best;
    long long best_fitness;
    std::vector<SamplePoint> history; // best-so-far sampled every SAMPLE_INTERVAL
    double elapsed_seconds;
};

static RunResult run_random_search()
{
    Chromosome best_ever;
    long long best_fitness_ever = 0;
    std::vector<SamplePoint> history;

    auto start_time = std::chrono::steady_clock::now();
    std::cout << "Pokretanje random pretrage...\n\n";

    for (long long evaluation = 1; evaluation <= EVALUATIONS; ++evaluation)
    {
        Chromosome candidate = init_random();
        long long f = fitness(candidate);
        if (f > best_fitness_ever)
        {
            best_fitness_ever = f;
            best_ever = candidate;
        }

        if (evaluation % SAMPLE_INTERVAL == 0 || evaluation == EVALUATIONS)
        {
            history.push_back({evaluation, best_fitness_ever});
        }
        if (evaluation % 100000 == 0 || evaluation == EVALUATIONS)
            std::cout << "Eval " << std::setw(8) << evaluation
                      << " | Fitness: " << std::setw(12) << f
                      << " | Best ever: " << best_fitness_ever << "\n";
    }

    auto end_time = std::chrono::steady_clock::now();
    double elapsed = std::chrono::duration<double>(end_time - start_time).count();

    return {std::move(best_ever), best_fitness_ever, std::move(history), elapsed};
}

// ── Results writing ──────────────────────────────────────────────────────────
static std::string timestamp_str()
{
    auto now = std::time(nullptr);
    char buffer[32];
    std::strftime(buffer, sizeof(buffer), "%Y-%m-%d_%H-%M-%S", std::localtime(&now));
    return buffer;
}

static void save_results(const std::vector<RunResult> &runs)
{
    fs::path base = fs::path(__FILE__).parent_path();
    fs::path results_dir = base / ("rezultati-" + PROBLEM_SIZE);
    fs::path txt_dir = results_dir / "rezultati";
    fs::create_directories(txt_dir);

    std::ostringstream name;
    name << "random-cpp-" << EVALUATIONS << "eval.txt";
    fs::path filepath = txt_dir / name.str();

    std::ofstream file(filepath);
    file << "Random Search Parametri (C++):\n"
         << "  Broj evaluacija po pokretanju: " << EVALUATIONS << "\n"
         << "  Broj pokretanja: " << NUM_RUNS << "\n\n";

    long long best_overall = 0, worst_overall = LLONG_MAX;
    double total_time = 0;
    std::vector<long long> profits;
    for (size_t i = 0; i < runs.size(); ++i)
    {
        file << "  Pokretanje " << (i + 1) << ": " << runs[i].best_fitness
             << " dinara  (" << std::fixed << std::setprecision(2) << runs[i].elapsed_seconds << "s)\n";
        profits.push_back(runs[i].best_fitness);
        best_overall = std::max(best_overall, runs[i].best_fitness);
        worst_overall = std::min(worst_overall, runs[i].best_fitness);
        total_time += runs[i].elapsed_seconds;
    }
    std::sort(profits.begin(), profits.end());
    double median = profits.size() % 2 ? (double)profits[profits.size() / 2]
                                       : (profits[profits.size() / 2 - 1] + profits[profits.size() / 2]) / 2.0;
    double mean = 0.0;
    for (long long p : profits)
        mean += (double)p;
    mean /= (double)profits.size();
    double variance = 0.0;
    for (long long p : profits)
        variance += ((double)p - mean) * ((double)p - mean);
    variance /= (double)profits.size();
    double stdev = std::sqrt(variance);

    file << "\n  PROSEK:   " << std::fixed << std::setprecision(1) << mean << "\n"
         << "  MEDIJANA: " << median << "\n"
         << "  NAJBOLJI: " << best_overall << "\n"
         << "  NAJGORI:  " << worst_overall << "\n"
         << "  STDEV:    " << stdev << "\n"
         << "  UKUPNO VREME: " << total_time << "s\n"
         << "  PROSECNO VREME: " << total_time / runs.size() << "s\n";

    std::cout << "\nRezultati sacuvani u: " << filepath << "\n";

    // Convergence CSV (same schema as ga-cpp-convergence)
    std::ostringstream csv_name;
    csv_name << "random-cpp-convergence-" << EVALUATIONS << "eval.csv";
    fs::path csv_dir = results_dir / "svi-podaci";
    fs::create_directories(csv_dir);
    fs::path csv_path = csv_dir / csv_name.str();
    std::ofstream csv(csv_path);
    csv << "run,generation,evaluations,best_fitness\n";
    // best_fitness = running best-so-far at the given evaluation count
    for (size_t i = 0; i < runs.size(); ++i)
        for (size_t g = 0; g < runs[i].history.size(); ++g)
            csv << (i + 1) << "," << g << ","
                << runs[i].history[g].evaluations << ","
                << runs[i].history[g].best_fitness << "\n";
    std::cout << "Konvergencija (CSV): " << csv_path << "\n";
}

// ── Entry Point ──────────────────────────────────────────────────────────────
int main()
{
    std::cout << "Ucitavanje podataka za " << PROBLEM_SIZE << "...\n";
    load_data();
    std::cout << "Dimenzije problema: " << NUM_SERVICES << " servisa x "
              << NUM_COMPUTERS << " racunara\n";

    std::cout << "\nRandom Search Parametri:\n"
              << "  Broj evaluacija po pokretanju: " << EVALUATIONS << "\n"
              << "  Uzorkovanje na svakih: " << SAMPLE_INTERVAL << " evaluacija\n";

    std::vector<RunResult> runs;
    for (int r = 0; r < NUM_RUNS; ++r)
    {
        std::cout << "\n===== Pokretanje " << (r + 1) << "/" << NUM_RUNS << " =====\n";
        runs.push_back(run_random_search());
        std::cout << "\nNajbolja zarada: " << runs.back().best_fitness
                  << " | Vreme: " << std::fixed << std::setprecision(2) << runs.back().elapsed_seconds << "s\n";
    }

    save_results(runs);
    return 0;
}
